clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -299,63 +356,51 @@ protected void onJoinComplete(int generation,
             this.nextAutoCommitTimer.updateAndReset(autoCommitIntervalMs);
 
         // execute the user's callback after rebalance
-        ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
+        final AtomicReference<Exception> firstException = new AtomicReference<>(null);
+        Set<TopicPartition> addedPartitions = new HashSet<>(assignedPartitions);
+        addedPartitions.removeAll(ownedPartitions);
 
+        // note that we should only change the assignment AFTER we've triggered the rebalance callback
```

Comment:
Why is that? I would have thought the callback should only be triggered after the partitions are actually assigned?

Reply:
I understand that for cooperative, we should call onPartitionsRevoked before actually revoking the partitions. But it seems like for eager we should first assign, then notify that the partitions have already been assigned through the callback. So maybe for cooperative we should update the assignment between `onPartitionsRevoked` and `onPartitionsAssigned`

Reply:
If we update the assignment first, then user trying to e.g. commit offset on `onPartitionRevoked` would get an exception saying it is not assigned anymore. So for either Eager or Cooperative, we need to make sure the partitions-going-to-be-revoked are still owned when the callback is triggered.

But semantically, yes, it is for notifications only so even if the callback throws some exception the assignment would still be updated afterwards.

Reply:
We're not calling `onPartitionsRevoked` here in the eager case, but we are changing the logic for eager (previously users could assume once `onPartitionsAssigned` was called the partitions were actually owned). 
I guess I'm hesitant to change this, since to me as a user I would interpret the `onPartitionsAssigned` callback as meaning the partitions are owned and ready to be fetched from/committed to/etc. Would this not introduce a race condition?

Reply:
Hmm, not sure I understand the latest comment. For eager case:

```
                // assign partitions that are not yet owned
                subscriptions.assignFromSubscribed(assignedPartitions);

                firstException.compareAndSet(null, invokePartitionsAssigned(addedPartitions));
```

we update the assignment before `invokePartitionsAssigned`.

In the old code:

```
if (!subscriptions.assignFromSubscribed(assignment.partitions())) {
```

is called before we proceed to the callback (note in the new code it was separated to the `checkAssignmentMatchedSubscription` and then the actual update. So it seems to me that the logic for EAGER does not change, right?

Reply:
Ah, sorry I didn't look closely enough -- you're right, the code in the eager case looks fine, it's just the comment that seems wrong. Can we edit it to match the actual behavior?

Reply:
Or possibly move it to the Cooperative case, and specify that we should only change after we've triggered the *onRevoked* callback -- as it is it's just incorrect

Reply:
I see :) yeah I've updated the comment per your other comment.